tracing01-xdp-simple/trace_load_and_stats.c

```diff
@@ -183,71 +181,37 @@ int filename__read_int(const char *filename, int *value)
 	return err;
 }
 
-#define TP "/sys/kernel/debug/tracing/events/"
-
-static int read_tp_id(const char *name, int *id)
-{
-	char path[PATH_MAX];
-
-	snprintf(path, PATH_MAX, TP "%s/id", name);
-	return filename__read_int(path, id);
-}
-
-static inline int
-sys_perf_event_open(struct perf_event_attr *attr,
-		    pid_t pid, int cpu, int group_fd,
-		    unsigned long flags)
-{
-	return syscall(__NR_perf_event_open, attr, pid, cpu, group_fd, flags);
-}
-
 static struct bpf_object* load_bpf_and_trace_attach(struct config *cfg)
 {
-	struct perf_event_attr attr;
 	struct bpf_object *obj;
-	int err, bpf_fd;
-	int id, fd;
-
-	if (bpf_prog_load(cfg->filename, BPF_PROG_TYPE_TRACEPOINT, &obj, &bpf_fd)) {
-		fprintf(stderr, "ERR: failed to load program\n");
-		goto err;
-	}
+	int err;
+	struct bpf_program *prog;
+	struct bpf_link *link;
 
-	if (read_tp_id("xdp/xdp_exception", &id)) {
-		fprintf(stderr, "ERR: can't get program section\n");
-		goto err;
-	}
+	obj = bpf_object__open_file(cfg->filename, NULL);
+	err = libbpf_get_error(obj);
+	if (err)
+			return NULL;
 
-	/* Setup tracepoint perf event */
-	memset(&attr, 0, sizeof(attr));
-	attr.type = PERF_TYPE_TRACEPOINT;
-	attr.config = id;
-	attr.sample_period = 1;
-
-	/* .. and open it */
-	fd = sys_perf_event_open(&attr, -1, 0, -1, 0);
-	if (fd <= 0) {
-		fprintf(stderr, "ERR: failed to open perf event %s\n",
-			strerror(errno));
-		goto err;
+	err = bpf_object__load(obj);
+	if (err) {
+			fprintf(stderr, "ERR: loading BPF-OBJ file(%s) (%d): %s\n",
+					cfg->filename, err, strerror(-err));
+			goto err;
 	}
 
-	/* Assign the program to the tracepoint perf event */
-	err = ioctl(fd, PERF_EVENT_IOC_SET_BPF, bpf_fd);
-	if (err) {
-		fprintf(stderr, "ERR: failed to connect perf event with eBPF prog %s\n",
-			strerror(errno));
-		goto err;
+	prog = bpf_object__next_program(obj, NULL);
+	if (!prog) {
+			fprintf(stderr, "ERR: Failed to retrieve program from BPF-OBJ file(%s) (%d): %s\n",
+					cfg->filename, err, strerror(-err));
+			goto err;
 	}
 
-	/* ... and finally enable the event. */
-	err = ioctl(fd, PERF_EVENT_IOC_ENABLE, 0);
-	if (err) {
-		fprintf(stderr, "ERR: failed to enable perf event %s\n",
-			strerror(errno));
+	link = bpf_program__attach_tracepoint(prog, "xdp", "xdp_exception");
```

Comment:
I think libbpf has auto-attach capability for `SEC("tracepoint/xdp/xdp_exception")` and not sure how that interacts with `bpf_program__attach_tracepoint()`.

Reply:
You mean that if I just load the prog some magic happens under the hood?

Reply:
Yes, although it's not well documented :-) so I'm not at all sure about how the behaviour gets triggered. It could just be libbpf-skeleton programs that auto-attach. This blog post only describes the negative case, disabling auto-attach: https://nakryiko.com/posts/libbpf-v1/ - search for "autoattach".

Reply:
Looking at the code, it maybe just that there is the generic `bpf_program__attach()` API that gets the attach point from the SEC definition: https://libbpf.readthedocs.io/en/latest/api.html#libbpf_8h_1a261653845c38fc86d6d7ed0746abfc12

Reply:
Yeah, so it's mostly a skeleton thing - my_skeleton__attach() will "auto-attach" all progs in the skeleton that are auto-attach capable and have declared the attach point in the SEC definition. https://liuhangbin.netlify.app/post/bpf-skeleton/

But the SEC definition also enables the generic `bpf_program__attach()` function.

So ignore me crying wolf, the above code will do the right thing. It's just overkill to use `bpf_program__attach_tracepoint()` _and_ declare `SEC("tracepoint/xdp/xdp_exception")`. `SEC("tracepoint") would have been enough. 

We should document all this somewhere 😆